README.rst

.. image:: https://circleci.com/gh/sdpython/onnxcustom/tree/master.svg?style=svg
    :target: https://circleci.com/gh/sdpython/onnxcustom/tree/master

.. image:: https://travis-ci.com/sdpython/onnxcustom.svg?branch=master
    :target: https://app.travis-ci.com/github/sdpython/onnxcustom
    :alt: Build status

.. image:: https://ci.appveyor.com/api/projects/status/a3sn45a2fayoxb5q?svg=true
    :target: https://ci.appveyor.com/project/sdpython/onnxcustom
    :alt: Build Status Windows

.. image:: https://codecov.io/gh/sdpython/onnxcustom/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/sdpython/onnxcustom

.. image:: https://badge.fury.io/py/onnxcustom.svg
    :target: http://badge.fury.io/py/onnxcustom

.. image:: http://img.shields.io/github/issues/sdpython/onnxcustom.png
    :alt: GitHub Issues
    :target: https://github.com/sdpython/onnxcustom/issues

.. image:: https://img.shields.io/badge/license-MIT-blue.svg
    :alt: MIT License
    :target: http://opensource.org/licenses/MIT

.. image:: https://pepy.tech/badge/onnxcustom/month
    :target: https://pepy.tech/project/onnxcustom/month
    :alt: Downloads

.. image:: https://img.shields.io/github/forks/sdpython/onnxcustom.svg
    :target: https://github.com/sdpython/onnxcustom/
    :alt: Forks

.. image:: https://img.shields.io/github/stars/sdpython/onnxcustom.svg
    :target: https://github.com/sdpython/onnxcustom/
    :alt: Stars

.. image:: https://img.shields.io/github/repo-size/sdpython/onnxcustom
    :target: https://github.com/sdpython/onnxcustom/
    :alt: size

onnxcustom: custom ONNX
=======================

.. image:: https://raw.githubusercontent.com/sdpython/onnxcustom/master/_doc/sphinxdoc/source/phdoc_static/project_ico.png
    :width: 50

`documentation <http://www.xavierdupre.fr/app/onnxcustom/helpsphinx/index.html>`_

Examples, tutorial on how to convert machine learned models into ONNX,
implement your own converter or runtime, or even train with ONNX / onnxruntime.

The function *check* or the command line ``python -m onnxcustom check``
checks the module is properly installed and returns processing
time for a couple of functions or simply:

::

    import onnxcustom
    onnxcustom.check()

The documentation also introduces *onnx*, *onnxruntime* for
inference and training.
The tutorial related to *scikit-learn*
has been merged into `sklearn-onnx documentation
<http://onnx.ai/sklearn-onnx/index_tutorial.html>`_.
Among the tools this package implements, you may find:

* a tool to convert NVidia Profilder logs into a dataframe,
* a SGD optimizer similar to what *scikit-learn* implements but
  based on *onnxruntime-training* and able to train an CPU and GPU,
* functions to manipulate *onnx* graph.

**Installation of onnxruntime-training**

onnxruntime-training is only available on Linux. The CPU
can be installed with the following instruction.

::

    pip install onnxruntime-training --extra-index-url https://download.onnxruntime.ai/onnxruntime_nightly_cpu.html

Versions using GPU with CUDA or ROCm are available. Check
`download.onnxruntime.ai <https://download.onnxruntime.ai/>`_
to find a specific version.
You can use it on Windows
inside WSL (Windows Linux Subsystem) or compile it for CPU:

::

    python tools\ci_build\build.py --skip_tests --build_dir .\build\Windows --config Release --build_shared_lib --build_wheel --numpy_version= --cmake_generator="Visual Studio 16 2019" --enable_training --enable_training_ops --enable_training_torch_interop

GPU versions work better on WSL, see `Build onnxruntime on WSL (Windows Linux Subsystem)
<http://www.xavierdupre.fr/app/onnxcustom/helpsphinx/blog/2021/2021-12-16_wsl.html>`_.
*onnxcustom* can be installed from pypi.


HISTORY.rst

.. _l-HISTORY:

=======
History
=======

current - 2022-05-30 - 0.00Mb
=============================

* #64: Fixes bug introduced by recent updates (2022-05-29)
* #62: Adds a notebook about convolution (2022-05-27)
* #63: Add GraphProto to the documentation (2022-04-28)
* #60: Extends notebook coverage (2022-03-07)
* #59: Removes unnecessary exceptions (2022-03-06)
* #58: Uses new API to retrieve gradient for a model (2022-03-03)
* #57: Fix opset for ai.onnx.ml in examples after updating to onnx==1.11 (2022-02-22)
* #56: Renames Z into Y_grad in loss functions returning grad (2022-02-14)
* #55: Removes unncessary nodes in onnx_derivative (2022-02-14)
* #54: Implements a function which returns the gradient (2022-02-13)

0.4.293 - 2022-02-11 - 0.07Mb
=============================

* #53: Implements get_trained_onnx to retrieve the trained model (2022-02-04)
* #51: Implements scoring functions (2022-02-02)
* #49: Check nan values during training (2022-01-30)

0.4.274 - 2022-01-29 - 0.06Mb
=============================

* #48: Checks gradients are identical to scikit-learn for neural networks (2022-01-25)
* #47: Compare gradient values with scikit-learn (2022-01-25)
* #45: Adds an example about classification (2022-01-23)
* #44: Uses bind_ortvalue_input instead of bind_input to be faster (2022-01-23)
* #32: Implements classificiation loss (2022-01-23)
* #43: Fixes the training of a binary classifier with weights (2022-01-22)
* #42: Implements binary log loss for first API of orttraining (2022-01-22)
* #41: Refactors to improve profiling analysis (fix penalty, export onnx graphs) (2022-01-15)
* #40: Improves training fwbw with caching. (2022-01-14)
* #37: Improves performance of caching (2022-01-09)
* #36: Makes sure all plt.show() have been disabled in examples (2022-01-04)
* #35: Reduces the number of calls to bind_ortvalue (2022-01-04)

0.3.245 - 2022-01-03 - 0.06Mb
=============================

* #34: Implements L1, L2 losses for partial training (2022-01-03)
* #31: Implements penalty when running the gradient (2022-01-03)
* #33: Implements more loss functions (2022-01-02)
* #30: Implements different learning rate strategies (2022-01-01)
* #26: Implements learning rate from neural network (2022-01-01)
* #29: Fixes #16, support weights while training (2021-12-30)
* #16: Support weights when training a model (2021-12-30)
* #28: Replaces OrtValue by C_OrtValue everywhere (2021-12-19)
* #24: Be more consistent with OrtValue, OrtDevice, C and python versions (2021-12-19)
* #27: Uses C_OrtDevice everywhere (2021-12-16)
* #15: Add example with TrainingAgent (error gradient outside) (2021-12-14)
* #25: Implements optimizers with forward, backward functionalities (2021-12-04)
* #21: Implements a mechanism that update training weights with SGDRegressor or MLPRegressor (2021-12-04)
* #23: Extend documentation (2021-12-01)
* #22: Move learning_rate logic in a separate class (2021-12-01)
* #20: Experiment markdown rendering (2021-12-01)
* #19: Implements training with forward, backward (2021-12-01)
* #18: Adds classes to train an ONNX gradient with TrainingAgent (2021-11-27)
* #17: Minimize the number of data copy while training a model (2021-11-26)
* #14: Optimize DataLoader to use iobinding (avoir copy) (2021-11-26)
* #13: Adds function plot_onnxs (2021-11-25)
* #12: Add more examples (2021-11-19)

0.2.122 - 2021-10-31 - 0.03Mb
=============================

0.2.117 - 2021-10-26 - 0.03Mb
=============================

* #11: Automates nvprof logs retrieval (2021-10-26)
* #10: Add parameter to evaluate the model on test data while training (2021-10-12)
* #9: Refactoring documentation (2021-10-08)
* #8: Add example to look into neural network on GPU (2021-10-07)
* #7: Refactoring (2021-10-04)
* #6: Add examples with orttraining (2021-10-04)
* #5: Fix examples, update documentation (2021-09-28)
* #4: Complex scenarios (2021-07-12)
* #3: Replaces OnnxSubOperator by OnnxSubEstimator (2021-03-31)

0.1.0 - 2020-07-09 - 0.04Mb
===========================

* #1: Add an example on black list, while list of operators. (2020-07-09)


examples/plot_benchmark_op.py
"""
.. _example-ort-training:

Benchmark an operator
=====================



.. contents::
    :local:

A simple example
++++++++++++++++

"""

import json
import numpy
from numpy.testing import assert_almost_equal
import pandas
from pandas import DataFrame
import matplotlib.pyplot as plt
from onnxruntime import InferenceSession, get_device, OrtValue, SessionOptions
from skl2onnx.common.data_types import FloatTensorType
from skl2onnx.algebra.onnx_ops import OnnxSlice, OnnxAdd, OnnxMul
from mlprodict.tools import measure_time
from tqdm import tqdm
from mlprodict.testing.experimental_c import code_optimisation
from mlprodict.onnxrt.ops_whole.session import OnnxWholeSession

print([code_optimisation(), get_device()])


###################################
# The functions to compare.

def build_ort_op(op_version=14, save=None, **kwargs):  # opset=13, 14, ...
    slices = kwargs['slices']
    slice1, slice2 = slices
    slice1 = slice(0, None) if slice1 is None else slice(*slice1)
    slice2 = slice(0, None) if slice2 is None else slice(*slice2)
    
    axes = []
    starts = []
    ends = []
    for i in [0, 1]:
        if slices[i] is None:
            continue
        axes.append(i)
        starts.append(slices[i][0])
        ends.append(slices[i][1])
    starts = numpy.array(starts, dtype=numpy.int64)
    ends = numpy.array(ends, dtype=numpy.int64)
    axes = numpy.array(axes, dtype=numpy.int64)
    node1 = OnnxSlice('X', starts, ends, axes, op_version=op_version)
    node2 = OnnxAdd(node1, numpy.array([1], dtype=numpy.float32),
                    op_version=op_version)
    node3 = OnnxSlice(node2, starts, ends, axes,
                      op_version=op_version)
    node4 = OnnxMul(node3, numpy.array([2], dtype=numpy.float32),
                    op_version=op_version, output_names=['Y'])
    onx = node4.to_onnx(inputs=[('X', FloatTensorType([None, None]))],
                        target_opset=op_version)
    sess = InferenceSession(onx.SerializeToString(),
                            providers=["CPUExecutionProvider"])
    if save is not None:
        with open(save, "wb") as f:
            f.write(onx.SerializeToString())

    def npy_fct(x):
        return ((x[slice1, slice2] + 1)[slice1, slice2] * 2).copy()
    
    rnd = numpy.random.randn(10, 10).astype(numpy.float32)
    expected = npy_fct(rnd)
    got = sess.run(None, {'X': rnd})[0]
    try:
        assert_almost_equal(expected, got)
    except AssertionError as e:
        raise AssertionError(
            "kwargs=%r slice1=%r slice2=%r shapes=%r ? %r "
            "(x[slice1, slice2].shape)=%r" % (
                kwargs, slice1, slice2, expected.shape,
                got.shape, rnd[slice1, slice2].shape)) from e

    if get_device() == 'GPU':
        sessg = InferenceSession(onx.SerializeToString(),
                                 providers=["CUDAExecutionProvider"])

        def run_gpu(x):
            io_binding = sessg.io_binding()
            io_binding.bind_input(
                name='X', device_type=x.device_name(), device_id=0,
                element_type=numpy.float32, shape=x.shape(),
                buffer_ptr=x.data_ptr())
            io_binding.bind_output('Y')
            return sessg.run_with_iobinding(io_binding)

        return onx, lambda x: sess.run(None, {'X': x}), npy_fct, run_gpu
    else:
        return onx, lambda x: sess.run(None, {'X': x}), npy_fct, None


###########################################
# The benchmark.


def loop_fct(fct, xs):
    for x in xs:
        fct(x)


def benchmark_op(repeat=10, number=10, name="Slice", shape_slice_fct=None,
                 save=None, opset=14, repeat_profile=1500, verbose=1):
    if verbose:
        print("[benchmark_op] start repeat=%d number=%d repeat_profile=%d"
              " opset=%d." % (repeat, number, repeat_profile, opset))
    res = []
    for dim in tqdm([8, 16, 32, 64, 100, 128, 200,
                     256, 400, 512, 600, 784, 800,
                     1000, 1024, 1200]):
        shape, slices = shape_slice_fct(dim)
        onx, ort_fct, npy_fct, ort_fct_gpu = build_ort_op(
            save=save, op_version=opset, slices=slices)
            
        n_arrays = 20
        if dim >= 512:
            n_arrays = 10 
        xs = [numpy.random.rand(*shape).astype(numpy.float32)
              for _ in range(n_arrays)]
        info = dict(shape=shape)

        ctx = dict(xs=xs, loop_fct=loop_fct)

        # numpy
        ctx['fct'] = npy_fct
        obs = measure_time(
            "loop_fct(fct, xs)",
            div_by_number=True, context=ctx, repeat=repeat, number=number)
        obs['dim'] = dim
        obs['fct'] = 'numpy'
        obs['shape'] = ",".join(map(str, shape))
        obs['slices'] = str(slices)
        obs.update(info)
        res.append(obs)

        # onnxruntime
        ctx['fct'] = ort_fct
        obs = measure_time(
            "loop_fct(fct, xs)",
            div_by_number=True, context=ctx, repeat=repeat, number=number)
        obs['dim'] = dim
        obs['fct'] = 'ort'
        obs['shape'] = ",".join(map(str, shape))
        obs['slices'] = str(slices)
        obs.update(info)
        res.append(obs)

        if ort_fct_gpu is not None:

            # onnxruntime
            ctx['xs'] = [
                OrtValue.ortvalue_from_numpy(
                    x, 'cuda', 0) for x in xs]
            ctx['fct'] = ort_fct_gpu
            obs = measure_time(
                "loop_fct(fct, xs)",
                div_by_number=True, context=ctx, repeat=repeat, number=number)
            obs['dim'] = dim
            obs['fct'] = 'ort_gpu'
            obs['shape'] = ",".join(map(str, shape))
            obs['slices'] = str(slices)
            obs.update(info)
            res.append(obs)

    # profiling CPU
    if verbose:
        print("[benchmark_op] done.")
        print("[benchmark_op] profile CPU.")
    so = SessionOptions()
    so.enable_profiling = True
    sess = InferenceSession(onx.SerializeToString(), so,
                            providers=["CPUExecutionProvider"])
    for i in range(0, repeat_profile):
        sess.run(None, {'X': xs[-1]}, )
    prof = sess.end_profiling()
    with open(prof, "r") as f:
        js = json.load(f)
    dfprof = DataFrame(OnnxWholeSession.process_profiling(js))
    dfprof['shape'] = ",".join(map(str, shape))
    dfprof['slices'] = str(slices)
    if verbose:
        print("[benchmark_op] done.")

    # profiling CPU
    if ort_fct_gpu is not None:
        if verbose:
            print("[benchmark_op] profile GPU.")
        so = SessionOptions()
        so.enable_profiling = True
        sess = InferenceSession(onx.SerializeToString(), so,
                                providers=["CUDAExecutionProvider"])
        for i in range(0, repeat_profile):
            x = ctx['xs'][-1]
            io_binding = sess.io_binding()
            io_binding.bind_input(
                name='X', device_type=x.device_name(), device_id=0,
                element_type=numpy.float32, shape=x.shape(),
                buffer_ptr=x.data_ptr())
            io_binding.bind_output('Y')
            sess.run_with_iobinding(io_binding)

        prof = sess.end_profiling()
        with open(prof, "r") as f:
            js = json.load(f)
        dfprofgpu = DataFrame(OnnxWholeSession.process_profiling(js))
        dfprofgpu['shape'] = ",".join(map(str, shape))
        dfprofgpu['slices'] = str(slices)
        if verbose:
            print("[benchmark_op] profile done.")
    else:
        dfprofgpu = None

    # Dataframes
    shape_name = str(shape).replace(str(dim), "N")
    df = pandas.DataFrame(res)
    piv = df.pivot('shape', 'fct', 'average')

    rs = piv.copy()
    for c in ['numpy', 'ort', 'ort_gpu']:
        if c in rs.columns:
            rs["numpy/%s" % c] = rs['numpy'] / rs[c]
    rs = rs[[c for c in rs.columns if "/numpy" not in c]].copy()

    # Graphs.
    fig, ax = plt.subplots(1, 2, figsize=(12, 4))
    piv.plot(logx=True, logy=True, ax=ax[0],
             title="%s benchmark\n%r"
                   " lower better" % (name, shape_name))
    ax[0].legend(prop={"size": 9})
    rs.plot(logx=True, logy=True, ax=ax[1],
            title="%s Speedup, baseline=numpy\n%r"
                  " higher better" % (name, shape_name))
    ax[1].plot([min(rs.index), max(rs.index)], [0.5, 0.5], 'g--')
    ax[1].plot([min(rs.index), max(rs.index)], [2., 2.], 'g--')
    ax[1].legend(prop={"size": 9})
    return dfprof, dfprofgpu, df, rs, ax


######################################
# The results.

nth = int(code_optimisation().split('=')[1])
cols_profile = ["shape", "slices", "args_op_name", 'args_provider']

##############################################
# shape = (100, N) - slice = [1:-1], :
# ++++++++++++++++++++++++++++++++++++

dfs = []
dfprof, dfprofgpu, df, piv, ax = benchmark_op(
    shape_slice_fct=lambda dim: ((256, dim), ((1, -1), None)),
    save="bslice.onnx", number=nth*4, repeat=8, repeat_profile=100*nth)

dfs.append(df)
piv2 = df.pivot("fct", "shape", "average")
print("slices = [1:-1], :")
print(piv.to_markdown())
print(dfprof.drop(['pid', 'tid', 'ts'], axis=1).groupby(
    cols_profile).sum().to_markdown())
if dfprofgpu is not None:
    print(dfprofgpu.drop(['pid', 'tid'], axis=1).groupby(
        cols_profile).sum().to_markdown())

##############################################
# shape = (100, N) - slice = :, [1:-1]
# ++++++++++++++++++++++++++++++++++++

dfs = []
dfprof, dfprofgpu, df, piv, ax = benchmark_op(
    shape_slice_fct=lambda dim: ((256, dim), (None, (1, -1))),
    save="bslice.onnx", number=nth*4, repeat=8, repeat_profile=100*nth)

dfs.append(df)
piv2 = df.pivot("fct", "shape", "average")
print("slices = :, [1:-1]")
print(piv.to_markdown())
print(dfprof.drop(['pid', 'tid', 'ts'], axis=1).groupby(
    cols_profile).sum().to_markdown())
if dfprofgpu is not None:
    print(dfprofgpu.drop(['pid', 'tid'], axis=1).groupby(
        cols_profile).sum().to_markdown())

##############################################
# shape = (100, N) - slice = [1:-1], [1:-1]
# ++++++++++++++++++++++++++++++++++++

dfs = []
dfprof, dfprofgpu, df, piv, ax = benchmark_op(
    shape_slice_fct=lambda dim: ((256, dim), ((1, -1), (1, -1))),
    save="bslice.onnx", number=nth*4, repeat=8, repeat_profile=100*nth)

dfs.append(df)
piv2 = df.pivot("fct", "shape", "average")
print("slices = [1:-1], [1:-1]")
print(piv.to_markdown())
print(dfprof.drop(['pid', 'tid', 'ts'], axis=1).groupby(
    cols_profile).sum().to_markdown())
if dfprofgpu is not None:
    print(dfprofgpu.drop(['pid', 'tid'], axis=1).groupby(
        cols_profile).sum().to_markdown())

